Wangscape/tilegen/CartesianPowerIterator.h

```diff
@@ -0,0 +1,161 @@
+{
+    for (size_t i = 0; i < mIterators.size(); i++)
+    {
+        ++mIterators[i];
+        if (i < mIterators.size() - 1)
+        {
+            if (mIterators[i] == getLast())
+            {
+                mIterators[i] = getFirst();
+                mValues[i] = *mIterators[i];
+            }
+            else
+            {
+                mValues[i] = *mIterators[i];
+                break;
+            }
+        }
+        else if (mIterators[i] != getLast())
+        {
+            mValues[i] = *mIterators[i];
+        }
+    }
+    return *this;
+}
+
+template<typename InputIt>
+inline bool CartesianPowerIterator<InputIt>::operator==(const CartesianPowerIterator& other) const
+{
+    if (getFirst() != other.getFirst())
+        return false; // different ranges
+    if (getLast() != other.getLast())
+        return false; // different ranges
+    if (getIterators().size() != other.getIterators().size())
+        return false; // different powers
+    if (*getIterators().crbegin() == getLast() &&
+        *other.getIterators().crbegin() == getLast())
+        return true; // both are cend
+    for (size_t i = 0; i < getIterators().size(); i++)
+        if (getIterator(i) != other.getIterator(i))
+            return false; // different phases
+    return true;
+}
+
+template<typename InputIt>
+inline bool CartesianPowerIterator<InputIt>::operator!=(const CartesianPowerIterator& other) const
+{
+    return !(*this == other);
+}
+
+template<typename InputIt>
+inline const typename CartesianPowerIterator<InputIt>::Values& CartesianPowerIterator<InputIt>::operator*() const
+{
+    return mValues;
+}
+
+template<typename InputIt>
+inline const typename InputIt::value_type& CartesianPowerIterator<InputIt>::get(size_t n) const
+{
+    return mValues[n];
+}
+
+template<typename InputIt>
+inline const InputIt& CartesianPowerIterator<InputIt>::getIterator(size_t n) const
+{
+    return mIterators[n];
+}
+
+template<typename InputIt>
+inline const typename CartesianPowerIterator<InputIt>::Iterators& CartesianPowerIterator<InputIt>::getIterators() const
+{
+    return mIterators;
+}
+
+template<typename InputIt>
+inline const typename InputIt::difference_type CartesianPowerIterator<InputIt>::coordinate(size_t n) const
+{
+    return std::distance(getFirst(), mIterators[n]);
+}
+
+template<typename InputIt>
+inline const InputIt& CartesianPowerIterator<InputIt>::getFirst() const
+{
+    return mFirst;
+}
+
+template<typename InputIt>
+inline const InputIt& CartesianPowerIterator<InputIt>::getLast() const
+{
+    return mLast;
+}
+
+template<typename InputIt>
+inline std::pair<size_t, size_t> CartesianPowerIterator<InputIt>::coordinates_2d() const
+{
+    size_t clique_size = std::distance(getFirst(), getLast());
+    size_t x = 0; size_t y = 0;
+    for (size_t i = 0; i < static_cast<size_t>(CORNERS); i++)
+    {
+        size_t& z = ((i + 1) % 2) ? x : y;
+        z *= clique_size;
+        z += coordinate(i);
+    }
```

Comment:
Oh, we've discussed it already. Now I think the best could be splitting this `for` into two `for` loops, one for `x` and one for `y`. And maybe a function for these lines:
```cpp
z *= clique_size;
z += coordinate(i);
```
What do you think? How we can make it easier to understand?